src/test/java/org/jenkinsci/plugins/workflow/cps/steps/SetStatusStepTest.java

```diff
@@ -0,0 +1,458 @@
+package org.jenkinsci.plugins.workflow.cps.steps;
+
+import hudson.model.Result;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import javax.annotation.Nonnull;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class SetStatusStepTest {
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Issue("JENKINS-43995")
+    @Test
+    public void setStatus() throws Exception {
+        WorkflowJob j = r.jenkins.createProject(WorkflowJob.class, "setStatus");
+        j.setDefinition(new CpsFlowDefinition("stage('outermost') {\n" +
+                "  stage('middle-unstable') {\n" +
+                "    echo 'hi there again'\n" +
+                "    status('SUCCESS')\n" +
+                "    stage('inner-unstable') {\n" +
+                "      status('UNSTABLE')\n" +
+                "    }\n" +
+                "  }\n" +
+                "  stage('middle-set-failure') {\n" +
+                "    echo 'hi there yet again'\n" +
+                "    stage('inner-set-failure') {\n" +
+                "      status('FAILURE')\n" +
+                "    }\n" +
+                "  }\n" +
+                "  stage('middle-thrown-error') {\n" +
+                "    echo('hi there')\n" +
+                "    stage('middle-inner-success') {\n" +
+                "      status('SUCCESS')\n" +
+                "    }\n" +
+                "    stage('middle-inner-nested-caught') {\n" +
+                "      echo 'pre-catch'\n" +
+                "      catchError {\n" +
+                "        withEnv(['FOO=BAR']) {\n" +
+                "          error('nested error')\n" +
+                "        }\n" +
+                "      }\n" +
+                "      echo 'post-catch'\n" +
+                "    }\n" +
+                "    stage('middle-inner-error') {\n" +
+                "      error('error')\n" +
+                "    }\n" +
+                "  }\n" +
+                "}\n", true));
+        WorkflowRun b = r.assertBuildStatus(Result.FAILURE, r.waitForCompletion(j.scheduleBuild2(0).waitForStart()));
+        /*
+Node dump follows, format:
+[ID]{parent,ids}(millisSinceStartOfRun) flowNodeClassName stepDisplayName [st=startId if a block end node]
```

Comment:
For the love of Azathoth, no - it's like reading a 🌵 .   Many, small testcases, please!   It's impossible to debug status problems with mega testcases, because the status issues tend to be complex. 

You may wish to have a more complex one that is only used for things like nested blocks/parallels.

Reply:
Well, that’s basically what these two tests are for - checking the more complex use cases. I do need to add simpler use case tests, though.

Reply:
Kept these two tests but renamed them to be more specific, along with adding a simple test, a test for missing a body, and a test for an unknown `Result` string.

Reply:
So, generally, what I've found is that these cases can be really complex to debug so it's extra important to keep basic tests simple and limit the intent of complex cases to the specific behavior you're trying to exercise -- see https://github.com/svanoort/pipeline-graph-analysis-plugin/blob/master/src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTimingTest.java -- the simpler cases are the only thing that made it sane. 

Bismuth in wf-api was a big nightmarish because of how nasty the complex parallel cases were.